### PR TITLE
Low effort no_std conversion

### PR DESCRIPTION
**Pull request by mriise**
Just a quick and dirty `no_std` conversion

- [x] tests need to be done in a different way.

x allow no_alloc using `&mut str` another time maybe, no-std should be done first

### CODE REVIEW

**Comment by PSeitz** [comment]
Nice, I considered of adding no_std, but I thought String wasn't supported there.

Tests should be fine when moved to the tests folder, since they are then compiled seperately.

with "no_alloc using &mut str", do you mean preallocate space in a String and pass it as parameter?

**Comment by mriise** [comment]
yep, but `&str` and `String` are quite different. `&str` come with lots of caveats but can be used in `no_std` 

A common pattern to supporting no_alloc env is to have the user give a mutable ref instead of trying to manage things in the lib. While it pushes more complexity to using side of things generally the people doing that sort of thing usually like more control over memory management anyway. also I was wrong about `&mut str`, it should be a mutable slice `&mut [u8]`.

our end of things might look something like
```rust
pub fn get_id_mut(output: &mut [u8]) -> Result<Ok, Err>
```

An alternative using the soon to be stabilized const generics could be a possibility, but as its currently unstable and barely an MVP it has a bit more quirks than is worth the effort but might be a good update later on.

**Comment by mriise** [comment]
this also looks like a viable option. https://docs.rs/arrayvec/0.5.2/arrayvec/struct.ArrayString.html then size of the backing array could be calculated by turning `max_size` into `const fn` and doing `ArrayString::<[u8; x_max_size()]>::new()`

**Comment by PSeitz** [comment]
Yes https://docs.rs/arrayvec/0.5.2/arrayvec/struct.ArrayString.html looks like an better option, because the method would also need to return the actual length of the string, which would be included in ArrayString
```
pub fn get_id_mut(output: &mut [u8]) -> Result<u32, Err>
```

Currently max_size depends on the which id function is called and is calculated during runtime. Not sure if it can be converted to a `const fn` currently. Could be also set manually, that would be little lame though :)

**Comment by mriise** [comment]
there we go, compile time max size!

**Comment by PSeitz** [comment]
Nice job! It should be const fn anyway and not calculated during runtime.
I like the refactoring, it's much cleaner like this

File: src/lib.rs [hunk 1]
```diff
@@ -4,224 +4,204 @@ This is a cool-id-generator.
...
+/// fn main() {
+///     let byte_array: [u8; Size::Short as usize] = [0u8; Size::Short as usize];
+/// }
+/// ```
+pub enum Size {
+    /// Creates ids in the format of {adjective}-{prefix}-{animal|job}-{name}
+    /// e.g. "unpleasant-steampunk-poet-gerald"
+    ///
+    /// Generates 1 billion combinations
+    Short = get_id_max_len() as isize,
```
Comment (PSeitz) [inline src/lib.rs]: Medium is probably more accurate, since the ids are not really short

File: src/lib.rs [hunk 2]
```diff
@@ -4,224 +4,204 @@ This is a cool-id-generator.
...
-        let prefix = get_animal_prefix(&mut rng);
-
-        // println!("Total Combinations {:?}", combinations_animals * ANIMAL_PREFIX.len());
-        format!("{}-the-{}-and-{}{}-{}", name, adj1, adj2, prefix, animal)
-    }else{
-        // It's a job!
-        let prefix = get_job_prefix(&mut rng);
-        // println!("Total Combinations {:?}", combinations_jobs * JOBS_PREFIX.len());
-        format!("{}-the-{}-and-{}{}-{}", name, adj1, adj2, prefix, job)
+    let animal_or_job = {
```
Comment (PSeitz) [inline src/lib.rs:35]: the intermediate string can be avoided I think, by using `std::fmt::Arguments` or something similiar

Comment (mriise) [reply src/lib.rs:35]: how does this look
```rust 
let animal_or_job:  (&&str, &&str) = {...}
...
format!("{}{}-{}-{}", adj1, animal_or_job.0, animal_or_job.1, name)
```

Comment (PSeitz) [reply src/lib.rs:35]: looks good, probably easier to understand than arguments

**Comment by PSeitz** [comment]
You can also add yourself under authors in Cargo.toml if you'd like

**Comment by mriise** [comment]
while having `const fn` is really cool, it does add a noticeable compile time penalty (which is quite annoying)

**Comment by PSeitz** [comment]
slow compile times, my nemesis
It is really cool, but also quite unnecessary to let it compute by the consumers instead of shipping it computed, since it does not change.

Maybe a rust build.rs script would be a better solution.

Edit: I just realized a build script would not change anything, it would still be executed when compiling the crate

**Comment by mriise** [comment]
While it would be run at compile time, you are able to say "only when the build.rs file is changed" which would be better than compiling every time users of the project compile.

A more ergonomic way would be to separate the list of words out to its own crate so it can build itself without unhygienically intruding on the lib source code since the only good way of reading the output from a build script is to use the `include!` macro

**Comment by PSeitz** [comment]
Could you explain the advantage of having a second crate? What would be in the second crate, only the words or also the build script?

**Comment by mriise** [comment]
a crate with all the words, enums of word types, and the build script. and since all the words we use are ASCII compatible, I'm thinking we could just encode all the words into byte arrays.

**Comment by PSeitz** [comment]
What would be the advantage of a second crate?

 `&'static str` is basically a byte array. Do you mean one large byte array, with generated offsets?

**Comment by mriise** [comment]
no_alloc should be in a seperate PR anyway.

File: src/lib.rs [hunk 3]
```diff
@@ -4,224 +4,174 @@ This is a cool-id-generator.
...
+/// }
+/// ```
+pub enum Size {
+    /// Creates ids in the format of {adjective}-{prefix}-{animal|job}-{name}
+    /// e.g. "unpleasant-steampunk-poet-gerald"
+    ///
+    /// Generates 1 billion combinations
+    Medium = get_id_max_len() as isize,
+    /// Creates ids in the format of {name}-the-{adjective}-and-{adjective}-prefix-{animal|job}
+    /// e.g. "unpleasant-steampunk-poet-gerald"
```
Comment (PSeitz) [inline src/lib.rs]: this example is from medium size

File: src/lib.rs [hunk 4]
```diff
@@ -4,224 +4,174 @@ This is a cool-id-generator.
...
+    ///
+    /// Generates 115 billion combinations
+    Long = get_long_id_max_len() as isize,
+    /// Creates ids in the format of {name1}-{name2}-the-{adjective}-and-{adjective}-prefix-{animal|job}
+    /// e.g. "unpleasant-steampunk-poet-gerald"
```
Comment (PSeitz) [inline src/lib.rs]: this example is from medium size

**Comment by PSeitz** [comment]
if you execute `git add -A .` it should be able to track moved files

File: src/words/job.rs [hunk 5]
```diff
@@ -1,5 +1,4 @@
-
-pub(crate) const JOBS: [&'static str;69] = [
+pub const JOBS: [&'static str; 69] = [
```
Comment (mriise) [inline src/words/job.rs:1]: This makes it public to everyone, this alright?

Comment (PSeitz) [reply src/words/job.rs:1]: sure that's fine

**Comment by PSeitz** [comment]
Published in crates.io with 1.0.0

**Pull request merged**